contrib: Add deterministic-unittest-coverage

maflcko opened the pull request:
The `contrib/devtools/test_deterministic_coverage.sh` script is problematic:

* It is written in bash. This can lead to issues when running with the ancient bash version shipped by macOS by default, or can lead to other compatibility issues, such as https://github.com/bitcoin/bitcoin/pull/31588#discussion_r1946784827. Also, pipefail isn't set, so IO errors may be silently ignored.
* It is based on gcov. This can lead to issues, such as https://github.com/bitcoin/bitcoin/pull/31588#pullrequestreview-2602169248 (possibly due to prefix-map), or https://github.com/bitcoin/bitcoin/pull/31588#issuecomment-2646395385 (gcovr processing error), or https://github.com/bitcoin/bitcoin/pull/31588#pullrequestreview-2605954001 (gcovr assertion error).
* The script is severely outdated, with the last update to `NON_DETERMINISTIC_TESTS` being in the prior decade.

Instead of patching around all issues one-by-one, just provide a fresh rewrite, based on the recently added `deterministic-fuzz-coverage` tool based on clang, llvm-cov, and llvm-profdata. (Initial feedback indicates that this is a more promising attempt: https://github.com/bitcoin/bitcoin/pull/31588#issuecomment-2649356408 and https://github.com/bitcoin/bitcoin/pull/31588#issuecomment-2649354598).

The new tool also sets `RANDOM_CTX_SEED=21` as suggested by hodlinator in https://github.com/bitcoin/bitcoin/pull/31588#issuecomment-2650784726.

Commit:
contrib: deterministic-fuzz-coverage fixups

* Name the fuzz_corpora dir after its real name.
* Add missing cargo lock file.
* Use git instead of diff command to increase compatibility
* Use --help instead of --version to increase compatibility
* Use assert consistently for unexpected errors.
* Remove redundant Stdio::from.
* Fix typos.

Commit:
gitignore: target/

Needed for the recently added
contrib/devtools/deterministic-fuzz-coverage/target/

Commit:
contrib: Add deterministic-unittest-coverage

This replaces the bash script with a tool based on clang/llvm tools.

dergoegge commented:
Concept ACK

theStack reviewed (commented):
Concept ACK

hodlinator reviewed (commented):
Concept ACK fafce12a4871ed66a868b460b779e794a89218eb

Commit:
doc: Remove unused -fPIC

This is harmless, but no longer needed after commit
06b9236f4327525875768af5fc554c651c5ec3cf.

hodlinator commented on contrib/devtools/deterministic-unittest-coverage/src/main.rs:
```diff
@@ -0,0 +1,124 @@
+    let profdata_file = build_dir.join("test_det_cov.profdata");
+    let run_single = |run_id: u8| {
+        let cov_txt_path = build_dir.join(format!("test_det_cov.show.{run_id}.txt"));
+        assert!(Command::new(test_exe)
+            .env("LLVM_PROFILE_FILE", &profraw_file)
+            .env("BOOST_TEST_RUN_FILTERS", filter)
+            .env("RANDOM_CTX_SEED", "21")
+            .status()
+            .expect("test failed")
+            .success());
```
> In any case, if this was changed, it should be applied to almost all `unwrap`,`expect`,or `panic` as well? For example, I could imagine that `expect("test failed")` could fail when the exe was missing the executable bit?

The `expect()`s here are fine since we've already run `sanity_check()` which has proper `exit(1)` behavior via `exit_help()`.

So I really do think since we are passing the user-input `filter` value, we should not be triggering an `assert!`. I think for the other commands where we are using fixed inputs, asserts are more okay.

hodlinator replied:
> No strong opinion, but I think this should either be done for all places, or for none.

Note that you are using `exit(1)` for some cases, so you are already drawing an arbitrary line.

A user using something like `miniscript` as input is more likely than an permission failure IMO. Agree that not compiling with coverage is somewhere in between though.

If we're going to introduce more Rust code in the project, and use it in an interactive way, I think we should spend some time to make the user experience on par or better than Bash in every way.

hodlinator commented on contrib/devtools/deterministic-unittest-coverage/src/main.rs:
```diff
@@ -0,0 +1,124 @@
+    let r0 = run_single(0);
+    let r1 = run_single(1);
+    check_diff(&r0, &r1);
+    println!("Coverage test passed.");
```
nit: Could provide a more satisfying message.
```suggestion
    println!("Coverage test passed - determinism held for given conditions in 2 consecutive runs!");
```

hodlinator reviewed (approved):
ACK fafa592a8ad95fcb64db3d342cdc7192ff7ef2cd

Was able to use this to diagnose and disable non-determinism issues for:
`cargo run --manifest-path ./contrib/devtools/deterministic-unittest-coverage/Cargo.toml -- $PWD/build util_tests
`

<details><summary>Hacky WIP pro-determinism patch resulting in pass</summary>

```diff
diff --git a/src/random.cpp b/src/random.cpp
index 5b605e988d..087894ffcd 100644
--- a/src/random.cpp
+++ b/src/random.cpp
@@ -441,6 +441,8 @@ public:
 
     ~RNGState() = default;
 
+    bool IsDeterministic() noexcept EXCLUSIVE_LOCKS_REQUIRED(!m_mutex) { LOCK(m_mutex); return m_deterministic_prng.has_value(); }
+
     void AddEvent(uint32_t event_info) noexcept EXCLUSIVE_LOCKS_REQUIRED(!m_events_mutex)
     {
         LOCK(m_events_mutex);
@@ -642,6 +644,7 @@ void ProcRand(unsigned char* out, int num, RNGLevel level, bool always_use_real_
     assert(num <= 32);
 
     CSHA512 hasher;
+    if (!rng.IsDeterministic() || always_use_real_rng) {
     switch (level) {
     case RNGLevel::FAST:
         SeedFast(hasher);
@@ -653,6 +656,7 @@ void ProcRand(unsigned char* out, int num, RNGLevel level, bool always_use_real_
         SeedPeriodic(hasher, rng);
         break;
     }
+    }
 
     // Combine with and update state
     if (!rng.MixExtract(out, num, std::move(hasher), false, always_use_real_rng)) {
@@ -771,7 +775,7 @@ FastRandomContext::FastRandomContext(bool fDeterministic) noexcept : requires_se
 void RandomInit()
 {
     // Invoke RNG code to trigger initialization (if not already performed)
-    ProcRand(nullptr, 0, RNGLevel::FAST, /*always_use_real_rng=*/true);
+    ProcRand(nullptr, 0, RNGLevel::FAST, /*always_use_real_rng=*/false);
 
     ReportHardwareRand();
 }
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index bf26997c07..a6349f3280 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -78,7 +78,6 @@ static FastRandomContext g_rng_temp_path;
 static const bool g_rng_temp_path_init{[] {
     // Must be initialized before any SeedRandomForTest
     Assert(!g_used_g_prng);
-    (void)g_rng_temp_path.rand64();
     g_used_g_prng = false;
     return true;
 }()};
```
</details>

Commit:
test: Exclude SeedStartup from coverage counts

hodlinator reviewed (approved):
re-ACK fa99c3b544b631cfe34d52fb5e71636aedb1b423

Minor text-only changes + Fix for source of differing coverage-counters.

I'm fine with or without the `ResetCoverageCounters()`-change. Only meant to demonstrate that this tool was useful by testing it out in my last ACK.

brunoerg reviewed (commented):
light ACK fa99c3b544b631cfe34d52fb5e71636aedb1b423

I have not reviewed the code yet, only checked that `deterministic-unittest-coverage` is working as expected. I tested on MacOS 14.3. 

```sh
cmake -B build_31901 -DCMAKE_CXX_FLAGS='-fprofile-instr-generate -fcoverage-mapping'
```
then I tested it for 3 tests:
```sh
RUST_BACKTRACE=1 cargo run --manifest-path ./contrib/devtools/deterministic-unittest-coverage/Cargo.toml -- $PWD/build_31901 coinselector_tests
RUST_BACKTRACE=1 cargo run --manifest-path ./contrib/devtools/deterministic-unittest-coverage/Cargo.toml -- $PWD/build_31901 addrman_tests
RUST_BACKTRACE=1 cargo run --manifest-path ./contrib/devtools/deterministic-unittest-coverage/Cargo.toml -- $PWD/build_31901 net_tests
```

All of them returned "The coverage was not deterministic between runs.". I noticed that the reason is basically locks and crypto functions as expected.

brunoerg commented:
> If you want to check a deterministic test case you can try util_string_tests or util_tests.

Both returned "The coverage was not deterministic between runs.". One of the differences:
```sh
 /Users/brunogarcia/projects/bitcoin-core-dev/src/crypto/sha512.h:
     1|       |// Copyright (c) 2014-2022 The Bitcoin Core developers
@@ -138933,7 +138933,7 @@
    62|      0|}
    63|       |
    64|       |inline int64_t GetPerformanceCounter() noexcept
-   65|    958|{
+   65|    966|{
    66|       |    // Read the hardware time stamp counter when available.
    67|       |    // See https://en.wikipedia.org/wiki/Time_Stamp_Counter for more information.
    68|       |#if defined(_MSC_VER) && (defined(_M_IX86) || defined(_M_X64))
@@ -138948,9 +138948,9 @@
    77|       |    return (r2 << 32) | r1;
    78|       |#else
    79|       |    // Fall back to using standard library clock (usually microsecond or nanosecond precision)
-   80|    958|    return std::chrono::high_resolution_clock::now().time_since_epoch().count();
-   81|    958|#endif
-   82|    958|}
+   80|    966|    return std::chrono::high_resolution_clock::now().time_since_epoch().count();
+   81|    966|#endif
+   82|    966|}
    83|       |
    84|       |#ifdef HAVE_GETCPUID
    85|       |bool g_rdrand_supported = false;
@@ -139185,22 +139185,22 @@
   314|       |    // Hash loop
   315|      2|    unsigned char buffer[64];
   316|      2|    const auto stop{SteadyClock::now() + dur};
-  317|    780|    do {
-  318|   780k|        for (int i = 0; i < 1000; ++i) {
-                                                ^780k
+  317|    788|    do {
+  318|   788k|        for (int i = 0; i < 1000; ++i) {
+                                                ^788k
   ------------------
-  |  Branch (318:25): [True: 780k, False: 780]
+  |  Branch (318:25): [True: 788k, False: 788]
   ------------------
-  319|   780k|            inner_hasher.Finalize(buffer);
-  320|   780k|            inner_hasher.Reset();
-  321|   780k|            inner_hasher.Write(buffer, sizeof(buffer));
-  322|   780k|        }
+  319|   788k|            inner_hasher.Finalize(buffer);
+  320|   788k|            inner_hasher.Reset();
+  321|   788k|            inner_hasher.Write(buffer, sizeof(buffer));
+  322|   788k|        }
   323|       |        // Benchmark operation and feed it into outer hasher.
-  324|    780|        int64_t perf = GetPerformanceCounter();
-  325|    780|        hasher.Write((const unsigned char*)&perf, sizeof(perf));
-  326|    780|    } while (SteadyClock::now() < stop);
+  324|    788|        int64_t perf = GetPerformanceCounter();
+  325|    788|        hasher.Write((const unsigned char*)&perf, sizeof(perf));
+  326|    788|    } while (SteadyClock::now() < stop);
   ------------------
-  |  Branch (326:14): [True: 778, False: 2]
+  |  Branch (326:14): [True: 786, False: 2]
   ------------------
   327|       |
   328|       |    // Produce output from inner state and feed it to outer hasher.
```

brunoerg commented:
> The last commit is supposed to fix that and it seemingly did for hodlinator and myself. Are you sure you compiled the right commit? Though, even if it persists, my recommendation would be to address it in a follow-up. The goal here is mostly to add the contrib script itself.

Yes, I just verified it is the right commit, I re-built it again and I'm getting the same. Anyway, I'm fine with the script, we can address it in a follow-up.

dergoegge reviewed (approved):
tACK fa99c3b544b631cfe34d52fb5e71636aedb1b423

janb84 reviewed (commented):
Concept ACK [fa99c3b](https://github.com/bitcoin/bitcoin/commit/fa99c3b544b631cfe34d52fb5e71636aedb1b423)

- Cloned & Build
- Runned the tooling, as instructed per Readme
- Reviewed the code

As a side note:
I have run into the same non-deterministic issues with util_string_tests or util_tests as [brunoerg](https://github.com/bitcoin/bitcoin/pull/31901#issuecomment-2684942423) although not important for this PR

Prabhat1308 commented:
Concept ACK [`fa99c3b`](https://github.com/bitcoin/bitcoin/pull/31901/commits/fa99c3b544b631cfe34d52fb5e71636aedb1b423)
Fully agree with removing the outdated bash script.

I have only checked the `deterministic-unittest-coverage` and not the `deterministic-fuzz-coverage` on MacOS 15.3.1 with clang 19 . It has been giving me some flaky results . On the first run both `util_string_tests` or `util_tests` were giving me result as deterministic coverage. 

<details>

<summary>First run</summary>

<img width="1035" alt="Screenshot 2025-03-01 at 7 58 12 PM" src="https://github.com/user-attachments/assets/5d0d6e85-c478-4377-a71c-e12f8942cd04" />

</details>

I ran them again and they started giving non-deterministic coverage
<details>

<summary>Second run</summary>

<img width="1055" alt="Screenshot 2025-03-01 at 7 59 38 PM" src="https://github.com/user-attachments/assets/3e7db7ab-07c7-4dd5-8632-8d94ecf337cd" />
<img width="1046" alt="Screenshot 2025-03-01 at 8 00 04 PM" src="https://github.com/user-attachments/assets/83f63e15-b191-4e30-8ffe-c52d76360a89" />
</details>

janb84 commented:
> @brunoerg @janb84 @Prabhat1308 I expect all of you were running into the issue on macOS? I probably can't fix it myself, but I'd investigate whether the last commit is working:
> 
> * Is your result different with or without the last commit?
> * Does `ResetCoverageCounters` work?
> * What is the coverage result for `src/test/util/setup_common.cpp`, especially `g_rng_temp_path_init`? (You can find it in `fuzz_det_cov.show.{run_id}.txt`)

In checking this issue I discovered that due to the change in [31161](https://github.com/bitcoin/bitcoin/pull/31161) the tooling is broken. (the path has changed from `src/test/` to `/bin` )

brunoerg commented:
> Word of caution, i'm running in a nix-shell on a mac. So would like to see someone on pure mac env. confirm that this is the issue

Confirmed. I still have the issue on master but I just tested with the following diff and seems to work:

```diff
diff --git a/src/test/util/coverage.cpp b/src/test/util/coverage.cpp
index bbf068a6fa..ac0fb00e36 100644
--- a/src/test/util/coverage.cpp
+++ b/src/test/util/coverage.cpp
@@ -4,7 +4,7 @@

 #include <test/util/coverage.h>

-#if defined(__clang__) && defined(__linux__)
+#if defined(__clang__)
 extern "C" void __llvm_profile_reset_counters(void) __attribute__((weak));
 extern "C" void __gcov_reset(void) __attribute__((weak));
```

```
➜  bitcoin-core-dev git:(master) ✗ RUST_BACKTRACE=1 cargo run --manifest-path ./contrib/devtools/deterministic-unittest-coverage/Cargo.toml -- $PWD/build util_string_tests
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.00s
     Running `contrib/devtools/deterministic-unittest-coverage/target/debug/deterministic-unittest-coverage /Users/brunogarcia/projects/bitcoin-core-dev/build_31901 util_string_tests`
Running 1 test case...

*** No errors detected
Running 1 test case...

*** No errors detected
The coverage was deterministic across two runs.
➜  bitcoin-core-dev git:(master) ✗ RUST_BACKTRACE=1 cargo run --manifest-path ./contrib/devtools/deterministic-unittest-coverage/Cargo.toml -- $PWD/build util_string_tests
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.00s
     Running `contrib/devtools/deterministic-unittest-coverage/target/debug/deterministic-unittest-coverage /Users/brunogarcia/projects/bitcoin-core-dev/build_31901 util_string_tests`
Running 1 test case...

*** No errors detected
Running 1 test case...

*** No errors detected
The coverage was deterministic across two runs.
```

Prabhat1308 commented:
>  Word of caution, i'm running in a nix-shell on a mac. So would like to see someone on pure mac env. confirm that this is the issue

Applied this change and I get deterministic results on all runs
```
#if defined(__clang__)  && (defined(__linux__) || defined(__APPLE__)) 
```
<img width="1512" alt="Screenshot 2025-03-13 at 9 31 40 PM" src="https://github.com/user-attachments/assets/fb622440-8634-47ba-b21a-54fc567803df" />

> Confirmed. I still have the issue on master but I just tested with the following diff and seems to work:

I tested on the master branch and I dont seem to have an issue . These are the commands I ran

```
cmake -B build -DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang" \ 
   -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++" \
   -DCMAKE_C_FLAGS="-fprofile-instr-generate -fcoverage-mapping" \
   -DCMAKE_CXX_FLAGS="-fprofile-instr-generate -fcoverage-mapping"

cmake --build build -j$(sysctl -n hw.ncpu) 
RUST_BACKTRACE=1 cargo run --manifest-path ./contrib/devtools/deterministic-unittest-coverage/Cargo.toml -- $PWD/build util_tests  
```
